xla/service/gpu/transforms/copy_fusion_test.cc

```diff
@@ -508,5 +506,54 @@ TEST_F(CopyFusionTest, CopyFusionWithFusionReturningTupleAndOtherUser) {
               GmockMatch(m::Tuple(m::Negate(), m::Negate(), m::Copy())));
 }
 
+TEST_F(CopyFusionTest, FusionFitsInBudget) {
+  auto module = ParseAndReturnVerifiedModule(absl::StrCat(kModulePrefix, R"(
+    fused_computation {
+      p1.1 = f32[128,512,28,28]{3,2,1,0} parameter(0)
+      mul = f32[128,512,28,28]{3,2,1,0} multiply(p1.1, p1.1)
+      ROOT neg = f32[128,512,28,28]{3,2,1,0} negate(mul)
+    }
+
+    ENTRY entry {
+      p0 = f32[128,512,28,28]{3,2,1,0} parameter(0)
+      fusion = f32[128,512,28,28]{3,2,1,0} fusion(p0), kind=kInput, calls=fused_computation
+      copy.1 = f32[128,512,28,28]{3,2,1,0} copy(fusion)
+      copy.2 = f32[128,512,28,28]{3,2,1,0} copy(fusion)
+      ROOT root = (f32[128,512,28,28]{3,2,1,0}, f32[128,512,28,28]{3,2,1,0}) tuple(copy.1, copy.2)
+    })"))
+                    .value();
+  ASSERT_TRUE(cf_.Run(module.get()).value());
+  SCOPED_TRACE(module->ToString());
+  const HloInstruction* root = module->entry_computation()->root_instruction();
+  const HloInstruction* fusion = nullptr;
+  ASSERT_THAT(root, GmockMatch(m::Tuple(m::GetTupleElement(m::Fusion(&fusion)),
+                                        m::GetTupleElement())));
+  EXPECT_THAT(fusion->fused_expression_root(),
+              GmockMatch(m::Tuple(m::Negate(), m::Copy(), m::Copy())));
+}
+
+TEST_F(CopyFusionTest, FusionExceedsBudget) {
+  auto module = ParseAndReturnVerifiedModule(absl::StrCat(kModulePrefix, R"(
+    fused_computation {
+      p1.1 = f32[128,128,128]{2,1,0} parameter(0)
```

Comment:
Hmm,  couple of things:

- Are you sure this test fails before your change? I would expect it to fail the OnlyElementwiseOpsReachableFromParams test, so there should be no copy fusion.
- This fusion doesn't use any shared memory (because none of the shared memory emitters support it), so it shouldn't actually exceed the budget. If it does, that's a bug.

Maybe you can create a fusion with many parameters with a loop. Something like this: https://github.com/openxla/xla/blob/159144edb8cc53ebc3bad4fea14737a5690dade3/xla/service/collective_ops_utils_test.cc#L155-L171. Your fusion kind should be `kLoop`, not `kInput`.